src/Grumplestiltskin/EllipticCurve.hs

```diff
@@ -0,0 +1,175 @@
+-- this function calculates the kP
+mulPoint :: Integer -> Point -> Point
+
+-- Vertical symmetry
+-- x axis stays the same, and y is inversed using modulo, i.e. P.y + inv(P).y == field modulus
+invPoint :: Point -> Point
+
+## Tests
+Some tests may not be feasible since we don't know the curve's order.
+
+Associativity `(aP + bP) + cP == aP + (bP + cP)`
+Inverse `P + inv(P) = Identity`
+`doublePoint(P) == P + P`
+`xP == (x + k)P` when k is the order of the curve
+`(a + b % curve_order)P == aP + bP` (since we don't know curve order we should pick sufficiently small numbers)
+`invPoint P == P` when P is identity (in case we represent identity in some special way)
+
+## Notes
+
+From the resource:
+> What about optimized_bn128?
+
+Examining the library, you will see an implementation called optimized_bn128.
+If you benchmark the execution time, you will see this version runs much quicker,
+and it is the implementation used by pyEvm. For educational purposes however,
+it is preferable to use the non-optimized version as it structures the points in a more intuitive way (the usual x, y tuple).
+The optimized version structures EC points as 3-tuples, which are harder to interpret.
+
+13.2 section in Handbook of Elliptic and Hyperelliptic Curve Cryptography -- Henri Cohen
+-}
+
+newtype PECPoint (s :: S) = PECPoint (Term s (PPair PGFElement PGFElement))
+    deriving stock
+        (Generic)
+    deriving anyclass
+        (SOP.Generic)
+    deriving
+        (PlutusType)
+        via (DeriveNewtypePlutusType PECPoint)
+
+instance PEq PECPoint
+
+newtype PECIntermediatePoint (s :: S) = PECIntermediatePoint (Term s (PPair PGFIntermediate PGFIntermediate))
+    deriving stock
+        (Generic)
+    deriving anyclass
+        (SOP.Generic)
+    deriving
+        (PlutusType)
+        via (DeriveNewtypePlutusType PECIntermediatePoint)
+
+pToPoint :: forall (s :: S). Term s PPositive -> Term s PECIntermediatePoint -> Term s PECPoint
+pToPoint fieldModulus p =
+    pcon $
+        PECPoint $
+            pmatch
+                (pupcast p)
+                (\(PPair x y) -> pcon $ PPair (pgfToElem x fieldModulus) (pgfToElem y fieldModulus))
+
+{- | @P + Q = R@ where @R@ is the inverse of a point on the intersection of the curve and the line defined by @P@ and @Q@.
+In case @P == Q@, we calculate the @2P@.
+-}
+pAddPoints :: forall (s :: S). Term s PPositive -> Term s PInteger -> Term s PECIntermediatePoint -> Term s PECIntermediatePoint -> Term s PECIntermediatePoint
+pAddPoints fieldModulus curveA point1 point2 =
+    pmatch
+        (pupcast point1)
+        ( \(PPair pointX1 pointY1) ->
+            pmatch
+                (pupcast point2)
+                ( \(PPair pointX2 pointY2) ->
+                    let
+                        -- TODO: maybe for the `yDiff` and `xDiff`, the `plet` should be used?
+                        yDiff = pointY1 #- pointY2
+                        xDiff = (pointX1 #- pointX2)
```

Comment:
If you end up using `yDiff` or `xDiff` multiple times in the computation, definitely `plet` them.